src/dotnet/OpenAPI.Validator/Validation/ArmResourcePropertiesBag.cs

```diff
@@ -47,19 +48,64 @@ public class ArmResourcePropertiesBag : TypedRule<Dictionary<string, Schema>>
+            foreach(string resourceModelName in resourceModelsWithPropertiesBag)
             {
-                var violatingProperties = definitions[violatingModel].Properties.Keys.Union(definitions[violatingModel].Properties["properties"].Properties.Keys).Intersect(ArmPropertiesBag);
-                yield return new ValidationMessage(new FileObjectPath(context.File, context.Path.AppendProperty(violatingModel).AppendProperty("properties")), this, violatingModel,
-                                                   string.Join(", ", violatingProperties));
+                Dictionary<string, IEnumerable<string>> violations = new Dictionary<string, IEnumerable<string>>();             
+                CheckModelForViolation(definitions[resourceModelName], resourceModelName, definitions, context, violations);
+
+                foreach (string modelName in violations.Keys)
+                {
+                    yield return new ValidationMessage(new FileObjectPath(context.File,
+                    context.Path.AppendProperty(modelName).AppendProperty("properties")), this, modelName,
+                                                   string.Join(", ", violations[modelName]));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Method to check the model for violation of this rule.
+        /// This method checks if the model has propeties bag. If so and it
+        /// has reference then the reference is processed. Else, the individual
+        /// properties are checked for violation.
+        ///
+        /// Then, if the model has allOf's, they need to be processed individually.
+        /// </summary>
+        /// <param name="resourceModel">schema of the model</param>
+        /// <param name="resourceModelName">name of the model</param>
+        /// <param name="definitions">definitions</param>
+        /// <param name="context">context</param>
+        /// <param name="violations">violations table</param>
+        private void CheckModelForViolation(Schema resourceModel, string resourceModelName, Dictionary<string, Schema> definitions, RuleContext context, Dictionary<string, IEnumerable<string>> violations)
+        {
+            if (resourceModel.Properties?.ContainsKey("properties") == true)
+            {
+                string referenceName = resourceModel.Properties["properties"].Reference;
+                if (!string.IsNullOrWhiteSpace(referenceName))
+                {
+                    CheckModelForViolation(Schema.FindReferencedSchema(referenceName, definitions), Validator.Extensions.StripDefinitionPath(referenceName), definitions, context, violations);
+                }
+                else
+                {
+                    IEnumerable<string> violatingProperties = resourceModel.Properties["properties"].Properties?.Keys?.Intersect(ArmPropertiesBag);
+                    if (violatingProperties?.Any() == true)
```

Comment:
Curious: Does `Intersect` ever return `null` ? I thought it doesn't. If it doesn't we can just write 

`if (violatingProperties.Any())`